examples/htool.c

```diff
@@ -251,6 +251,413 @@ static int do_target_reset(uint32_t reset_option) {
+
+  while (left < right) {
+    uint32_t mid = (left+right) / 2;
+    if (command_payload_read(dev, mid, 1, buf) != 0) {
+      right = mid;
+    } else {
+      left = mid + 1;
+    }
+  }
+
+  DONE:
+  printf("Payload size is %u\n", right);
+}
+
+static int command_payload_update(const struct htool_invocation* inv) {
+  struct libhoth_usb_device* dev = htool_libhoth_usb_device();
+  if (!dev) {
+    return -1;
+  }
+
+  bool dry_run = false;
+  char* x = getenv("DRY_RUN");
+  if (x) {
+    dry_run = atoi(x);
+  }
+
+  int result = 0;
+  struct {
+    const char* source_file;
+  } args;
+  if (htool_get_param_string(inv, "source-file", &args.source_file)) {
+    return -1;
+  }
+
+  printf("1. Sending Initiate\n");
+  if (!dry_run) {
+    result = send_payload_update_command(dev, 0, 0, 0);
+    if (result != 0) {
+      printf("sending INITIATE failed.\n");
+      return -1;
+    }
+  }
+
+  printf("source_file: %s\n", args.source_file);
+
+  int fd = open(args.source_file, O_RDONLY, 0);
+  if (fd == -1) {
+    fprintf(stderr, "Error opening file %s: %s\n", args.source_file,
+            strerror(errno));
+    return -1;
+  }
+
+  const size_t buf_size = sizeof(struct payload_update_packet) + 2048;
+  uint8_t buffer[buf_size];
+
+  struct stat statbuf;
+  if (fstat(fd, &statbuf)) {
+    fprintf(stderr, "fstat error: %s\n", strerror(errno));
+    goto cleanup1;
+  }
+  if (statbuf.st_size > SIZE_MAX) {
+    fprintf(stderr, "file to large\n");
+    goto cleanup1;
+  }
+
+  size_t file_size = statbuf.st_size;
+  printf("File size: %zd\n", file_size);
+  uint8_t* file_data = mmap(NULL, file_size, PROT_READ, MAP_PRIVATE, fd, 0);
+
+  printf("2. Sending Data\n");
+  const size_t max_packet_size = 1024 - sizeof(struct payload_update_packet) -
+                                 sizeof(struct ec_host_request);
+
+  struct payload_update_packet* request1 = &buffer;
+  uint8_t* request1_payload = &buffer[sizeof(struct payload_update_packet)];
+  size_t rlen = 0;
+  uint8_t response;
+  int idx = 0;
+  int offset1 = 0;
+  int size = file_size;
+  uint8_t* data = file_data;
+  while (size > 0) {
+    size_t skip = size_to_skip(data, size);
+    data += skip;
+    size -= skip;
+    offset1 += skip;
+    if (size == 0) {
+      break;
+    }
+    size_t max_size_to_send = (size < max_packet_size ? size : max_packet_size);
+    size_t to_send = size_to_send(data, max_size_to_send);
+    if (to_send) {
+      request1->offset = offset1;
+      request1->len = to_send;
+      request1->type = 1;  // PAYLOAD_UPDATE_CONTINUE
+      memcpy(request1_payload, data, to_send);
+      result = 0;
+    RETRY:
+      if (!dry_run) {
+        result = htool_exec_hostcmd(
+            dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_PAYLOAD_UPDATE,
+            NULL, request1, sizeof(*request1) + request1->len, &response,
+            sizeof(response), &rlen);
+      }
+      if (result != 0) {
+        printf("Error, result=%d. will retry\n", result);
+        goto RETRY;
```

Comment:
This goto can be a loop:

```
while (1) {
  result = htool_exec_hostcmd(...);
  if (result == 0) {
    break;
  }
  printf("Error, ...\n");
}
```

But also, this retries indefinitely, and probably should give up eventually (after some number of tries, or a timeout).  Might be better to refactor this to a separate function, `htool_exec_hostcmd_with_retries` or similar.

Reply:
Thanks for the comments, will address this in a separate change focused on payload update.